### PR TITLE
Aszlam/memory updates

### PR DESCRIPTION
**Pull request by aszlam**
# Description

Please include a summary of the change and link to an issue this PR fixes. Please also include relevant context. List any downstream components that might be affected by this change.

Fixes # (issue)

## Type of change

Please check the options that are relevant.

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

This PR fixes a longstanding bug where updates to certain memories did not change the updated_time in the main Memories table.  Also some memory types (in particular Triples) were not tracked by the main Memories table-  as of this PR they all should be.  Finally, to support these changes, a db_write() method wrapping ._db_write() is implemented, and several raw SQL deletes are removed from code.

# Testing

A basic memory test is added in base agent

# Checklist:

- [ ] I have performed manual end-to-end testing of the feature in my environment.
- [x] I have added Docstrings and comments to the code.
- [x] I have made changes to existing documentation where needed.
- [x] I have added tests that show that the PR is functional.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have added relevant collaborators to review the PR before merge.

### CODE REVIEW

File: base_agent/base_memory_schema.sql [hunk 1]
```diff
@@ -87,11 +105,20 @@ CREATE TABLE Tasks (
 );
 CREATE INDEX TasksFinishedAt ON Tasks(finished_at);
 
+CREATE TRIGGER tasks_update AFTER UPDATE ON Tasks
```
Comment (snyxan) [inline base_agent/base_memory_schema.sql:108]: to keep consistent: tasks_update -> TasksUpdate

File: base_agent/base_memory_schema.sql [hunk 2]
```diff
@@ -107,13 +134,19 @@ CREATE TABLE Triples (
...
+    FOREIGN KEY(uuid) REFERENCES Memories(uuid) ON DELETE CASCADE,
+    FOREIGN KEY(subj) REFERENCES Memories(uuid) ON DELETE CASCADE,
+    FOREIGN KEY(obj) REFERENCES Memories(uuid) ON DELETE CASCADE
 );
 CREATE INDEX TriplesSubjPred ON Triples(subj, pred);
 CREATE INDEX TriplesPredObj ON Triples(pred, obj);
 CREATE INDEX TriplesSubjPredText ON Triples(subj, pred_text);
 CREATE INDEX TriplesPredTextObjText ON Triples(pred_text, obj_text);
 
+CREATE TRIGGER triples_update AFTER UPDATE ON Triples
```
Comment (snyxan) [inline base_agent/base_memory_schema.sql:146]: triples_update -> TriplesUpdate

File: base_agent/memory_nodes.py [hunk 3]
```diff
@@ -205,7 +205,100 @@ def create(cls, memory, name, snapshot=False) -> str:
...
+        )
+
+    @classmethod
+    def create(
+        cls,
+        memory,
+        snapshot: bool = False,
+        subj: str = "",  # this is a memid if given
+        obj: str = "",  # this is a memid if given
+        subj_text: str = "",
```
Comment (snyxan) [inline base_agent/memory_nodes.py:257]: probably use:
```
subj_text: str = None,
obj_text: str = None,
```

File: base_agent/memory_nodes.py [hunk 4]
```diff
@@ -205,7 +205,100 @@ def create(cls, memory, name, snapshot=False) -> str:
...
+        assert not (subj and subj_text)
+        assert not (obj and obj_text)
+        # TODO check if triple exists, don't make it again
+        memid = cls.new(memory, snapshot=snapshot)
+        pred = NamedAbstractionNode.create(memory, pred_text)
+        if not obj:
+            obj = NamedAbstractionNode.create(memory, obj_text)
+        if not subj:
+            subj = NamedAbstractionNode.create(memory, subj_text)
+        if not subj_text:
```
Comment (snyxan) [inline base_agent/memory_nodes.py:287]: and delete this?

File: base_agent/sql_memory.py [hunk 5]
```diff
@@ -342,7 +363,6 @@ def basic_search(self, filter_dict):
     ###  Triples  ###
     #################
 
-    # TODO should add a MemoryNode and a .create()
     def add_triple(
```
Comment (snyxan) [inline base_agent/sql_memory.py:366]: same as above

File: base_agent/sql_memory.py [hunk 6]
```diff
@@ -375,30 +395,14 @@ def add_triple(
...
+            self,
+            subj=subj,
+            obj=obj,
+            subj_text=subj_text,
+            pred_text=pred_text,
+            obj_text=obj_text,
+            confidence=confidence,
         )
 
     def tag(self, subj_memid: str, tag_text: str):
```
Comment (snyxan) [inline base_agent/sql_memory.py:408]: Returns:
int: number of rows affected -> memid of triple

File: base_agent/sql_memory.py [hunk 7]
```diff
@@ -416,8 +420,9 @@ def tag(self, subj_memid: str, tag_text: str):
             >>> tag_text = "shiny"
             >>> tag(subj_memid, tag_text)
         """
-        self.add_triple(subj=subj_memid, pred_text="has_tag", obj_text=tag_text)
+        return self.add_triple(subj=subj_memid, pred_text="has_tag", obj_text=tag_text)
 
+    # TODO remove_triple
     def untag(self, subj_memid: str, tag_text: str):
         """Delete tag for subject
```
Comment (snyxan) [inline base_agent/sql_memory.py:428]: We should either remove `Returns` in docstring or return the number of rows affected.
(in a separate PR maybe)

File: craftassist/agent/mc_memory_schema.sql [hunk 8]
```diff
@@ -74,6 +74,14 @@ CREATE TRIGGER VoxelObjectsUpdate AFTER UPDATE ON VoxelObjects
     WHEN ((SELECT COUNT(*) FROM VoxelObjects WHERE uuid=OLD.uuid LIMIT 1) == 0)
     BEGIN DELETE FROM Memories WHERE uuid=OLD.uuid;
 END; -- delete memory when last block is removed
+CREATE TRIGGER VoxelObjectsChanged AFTER UPDATE ON VoxelObjects
```
Comment (snyxan) [inline craftassist/agent/mc_memory_schema.sql:77]: VoxelObjectsChanged -> VoxelObjectsUpdate

File: craftassist/agent/mc_memory_schema.sql [hunk 9]
```diff
@@ -130,8 +138,18 @@ CREATE TABLE Mobs (
     FOREIGN KEY(uuid) REFERENCES Memories(uuid) ON DELETE CASCADE
 );
 
+CREATE TRIGGER MobsChanged AFTER UPDATE ON Mobs
```
Comment (snyxan) [inline craftassist/agent/mc_memory_schema.sql:141]: MobsChanged -> MobsUpdate

File: craftassist/agent/mc_memory_schema.sql [hunk 10]
```diff
@@ -130,8 +138,18 @@ CREATE TABLE Mobs (
...
 CREATE TABLE Rewards (
     uuid    NCHAR(36)       PRIMARY KEY,
     value   VARCHAR(32)     NOT NULL, -- {POSITIVE, NEGATIVE}
-    time    INTEGER         NOT NULL
+    time    INTEGER         NOT NULL,
+    
+    FOREIGN KEY(uuid) REFERENCES Memories(uuid) ON DELETE CASCADE
 );
+
+CREATE TRIGGER RewardsChanged AFTER UPDATE ON Rewards
```
Comment (snyxan) [inline craftassist/agent/mc_memory_schema.sql:153]: RewardsChanged -> RewardsUpdate

**Comment by snyxan** [comment]
For `low_level_memory_test.py`(and more files to come). This should be set up as unittest and added to CI.
In order to do so, we should add `__init__.py` under `test/` and have a `test.sh` file

**Comment by snyxan** [comment]
A bit unrelated: since we are replacing `_db_write` with `db_write`, we should do the same for all `_*` functions which are being used as public fn.

File: base_agent/sql_memory.py [hunk 11]
```diff
@@ -1154,3 +1162,19 @@ def safe_unpickle(self, bs):
                     delattr(obj, "__had_attr_" + attr)
                     setattr(obj, attr, self._safe_pickle_saved_attrs[obj.pickled_attrs_id][attr])
         return obj
+
+
+if __name__ == "__main__":
+
+    class P:
```
Comment (kavyasrinet) [inline base_agent/sql_memory.py:1169]: Is this being used somewhere ? or was it for local testing ? :) If the latter, we should remove

**Review by kavyasrinet — COMMENTED** [review]
Thanks for the PR! Added some comments too.

File: base_agent/memory_nodes.py (hunk 4 above)
Comment (aszlam) [reply base_agent/memory_nodes.py:287]: why delete these?

File: base_agent/sql_memory.py [hunk 12]
```diff
@@ -433,11 +438,14 @@ def untag(self, subj_memid: str, tag_text: str):
...
         """
-        self._db_write(
-            'DELETE FROM Triples WHERE subj=? AND pred_text="has_tag" AND obj_text=?',
+        # FIXME replace me with a basic filters when _self handled better
+        triple_memids = self._db_read(
+            'SELECT uuid FROM Triples WHERE subj=? AND pred_text="has_tag" AND obj_text=?',
             subj_memid,
             tag_text,
         )
+        if triple_memids:
```
[the remark this answers is not in the document]
Comment (aszlam) [reply base_agent/sql_memory.py:447]: the deletes are hard by default (so this is a hard delete).  as mentioned above, I can get rid of soft delete entirely if you think I should

File: base_agent/memory_nodes.py (hunk 4 above)
Comment (aszlam) [reply base_agent/memory_nodes.py:287]: oh lol, you mean 287-290 again, sorry

File: base_agent/sql_memory.py [hunk 13]
```diff
@@ -314,12 +320,27 @@ def forget(self, memid: str, hard=True):
...
             # eventually we might have second-order relations etc, this could set
             # off a chain reaction
-            self.remove_memid_triple(memid, role="both")
+
+    def forget_by_query(self, query: str, hard=True):
+        """remove memories from DB that match a query
+
+        Args:
+            query (string): should be "SELECT uuid FROM ... "
+            hard (bool): flag indicating whether it is a hard delete
```
[the remark this answers is not in the document]
Comment (aszlam) [reply base_agent/sql_memory.py:335]: maybe eventually we want to "stage" something for deletion, but not delete yet.  am happy to torch it for now, we aren't using soft deletes anywhere

[the remark this answers is not in the document]
Comment (kavyasrinet) [reply base_agent/sql_memory.py:335]: Great! Sounds good. Then yes, I vote for getting rid of soft delete altogether

**Review by snyxan — APPROVED** [review]

**Comment by snyxan** [comment]
I'll setup base_agent test in another PR.

**Pull request merged**